Capture and log all Python warnings

SunilMohanAdapa opened the pull request:
- Capture all Python warnings so that they can shown as part of logging system
  on console and in log file.

- Also capture deprecation warnings into logging system if debug mode is
  enabled. Current versions of Python disable deprecation warnings by default.
  Django 1.11 also follows this approach now.

Signed-off-by: Sunil Mohan Adapa <sunil@medhas.org>

JoKeyser commented:
This seems really useful, but sorry - how do you know it's working?
I tried to enable it, by adding `debug = True` into `/etc/plinth/plinth.config`... with no obvious effect (after restarting plinth). I've clicked around, but see no "debug" messages or "warnings" anywhere.
If that's normal - maybe would be good to log one "DEBUG MODE" message, to show it's enabled?

SunilMohanAdapa commented:
@JoKeyser, I ran `./run --debug` (my usual way or running Plinth during development) and I see at least one deprecation warning in the console messages related to new style MIDDLEWARE setting. I wonder why `debug=True` didn't have any affect. Perhaps that is a separate bug.

JoKeyser commented:
Ah, thanks, with `--debug` can confirm the warnings, great.

The `debug=True` in ` plinth.config` doesn't work because `config_items` doesn't contain the `debug` option, see in https://github.com/freedombox/Plinth/blob/5403d00e8579b3401ce2cbce94648ca4c8d9c02d/plinth/cfg.py#L79

One last question from me: https://docs.python.org/2/library/warnings.html lists also `ImportWarning`s are ignored by default; maybe they should also be enabled?

JoKeyser commented on plinth/__main__.py:
```diff
@@ -82,6 +83,14 @@ def setup_logging():
+    if cfg.debug:
+        warnings.filterwarnings('default', '', DeprecationWarning)
+        warnings.filterwarnings('default', '', PendingDeprecationWarning)
+
```
One last question from me: https://docs.python.org/3/library/warnings.html lists also ImportWarnings are ignored by default; maybe they should also be enabled?

SunilMohanAdapa commented:
Thank you for reviewing my patch and for being thorough.  No reason not to have ImportWarnings also, I added that change and tested by adding `warnings.warn('import warning', ImportWarning)`.  Rebased and push the change.

JoKeyser commented:
Okay great, glad I was helpful. I rebased and pushed, see a9c7e28, hence closing here.
